[quant][core][gpu] Implemented max pooling 2D using cudnn

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74673

Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

dzdang committed:
[quant][core][gpu] Implemented max pooling 2D using cudnn

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"




[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35135200).

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35135200).

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35135200).

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35135200).

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35135200).

jerryzh168 commented:
are we planning to register this under native_functions.yaml instead of a new quantized::maxpool function?

dzdang commented:
@jerryzh168 

> are we planning to register this under native_functions.yaml instead of a new quantized::maxpool function?

I think that's what we're trending towards with the other max pool PRs. I'll make that change

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Pooling.cpp:
```diff
@@ -0,0 +1,208 @@
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  check_maxpool2d_params(
```
so the dimension checks etc. can be removed after we have structured kernel support right? can we list the blockers for using that and add a TODO so that we can fix this later?

dzdang replied:
I think the structured kernel support for dimension checking is already there (just like it was there for the quantized cpu max pool 2d), but the blocker here is cudnn doesn't allow us to obtain indices for max pool2d.
so I don't think there's a way to implement max_pool2d_with_indices_out_quantized_cuda unless we want to return nothing for indices?

jerryzh168 replied:
oh I see, maybe we can add this context in the comment

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -1189,6 +1189,51 @@ def test_max_pool1d(self, X, kernel, stride, dilation, padding, ceil_mode):
+    @unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
+                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
+                   "after it is built by default")
+    def test_max_pool2d_cudnn(self, X, kernel, stride, dilation, padding, ceil_mode):
```
we can also add a TODO to merge this with the other maxpool test after CUDNN_V8 flag is enabled

jerryzh168 reviewed (approved):
looks good!

dzdang committed:
Update on "[quant][core][gpu] Implemented max pooling 2D using cudnn"


Summary:
Quantized 2D max pooling was implemented using cudnn. A corresponding
test case was also added. The operator requires cudnn version 8.3.3
or higher. v7 APIs (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward)
are currently used as there are currently no v8
APIs for pooling. Note the current implementation does not support dilated pooling
as it is not supported in cudnn.

Test plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_max_pool2d_cudnn
```

Differential Revision: [D35135200](https://our.internmc.facebook.com/intern/diff/D35135200)

[ghstack-poisoned]
